fix(compiler-cli): Use typescript to resolve modules for metadata generation

jeffora opened the pull request:
The current module resolution for the metadata collector simply attaches `.ts` to the import/export path, which does not work if the path is using Node / CommonJS behavior to resolve to an index.ts file. This patch uses typescript's module resolution logic, and will attempt to load the original typescript file if this resolution returns a `.js` or `.d.ts` file

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

The metadata collector does not generate correct metadata when a file uses an implied "index" in an import or export path. In the following example, the `import` refers to a path `../tools` that will be resolved by typescript as `../tools/index.ts`. The metadata collector, however, simply appends `.ts` directly to the path string before attempting to find it to generate metadata, thus looking for `../tools.ts`, which doesn't exist. This results in the metadata having no information for named entry `ɵb`, despite having it listed in imports and origins.

```ts
// app.component.ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { ToolModule } from '../tools';
import { AppComponent } from './app.component';


@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    ToolModule,
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

```ts
// ../tools/index.ts
import { NgModule } from '@angular/core';

@NgModule({
  declarations: [],
  exports: [],
  imports: [],
  providers: [],
})
export class ToolModule {

}
```

```js
{
  "__symbolic": "module",
  "version": 4,
  "metadata": {
    "AppModule": {
      "__symbolic": "class",
      "decorators": [
        {
          "__symbolic": "call",
          ...
          "arguments": [
            {
              ...
              "imports": [
                {
                  "__symbolic": "reference",
                  "module": "@angular/platform-browser",
                  "name": "BrowserModule",
                  "line": 13,
                  "character": 4
                },
                {
                  "__symbolic": "reference",
                  "name": "ɵb"
                }
              ],
              ...
            }
          ]
        }
      ],
      "members": {}
    },
    "ɵa": {
      "__symbolic": "class",
      ...
    }
  },
  "origins": {
    "AppModule": "./src/app/app.module",
    "ɵa": "./src/app/app.component",
    "ɵb": "./src/tools"
  },
  ...
}
```

Issue Number: N/A


## What is the new behavior?

This PR introduces a change to use the typescript module resolver to find files for metadata instead of just appending `.ts` to the input path name. This has the benefit that it will follow the same conventions that the runtime module resolution will use, and supports correctly finding files in all existing situations, as well as all situations where the path actually points to an index file (i.e. `../tools` will find `../tools.ts` or `../tools/index.ts` depending on how the project is structured).

With this change, correct metadata is generated for the same project as defined above - as seen below, `ɵb` now has metadata properly generated:

```js
{
  "__symbolic": "module",
  "version": 4,
  "metadata": {
    "AppModule": {
      "__symbolic": "class",
      "decorators": [
        {
          "__symbolic": "call",
          ...
          "arguments": [
            {
              ...
              "imports": [
                {
                  "__symbolic": "reference",
                  "module": "@angular/platform-browser",
                  "name": "BrowserModule",
                  "line": 13,
                  "character": 4
                },
                {
                  "__symbolic": "reference",
                  "name": "ɵb"
                }
              ],
              ...
            }
          ]
        }
      ],
      "members": {}
    },
    "ɵa": {
      "__symbolic": "class",
      ...
    },
    "ɵb": {
      "__symbolic": "class",
      "decorators": [
        {
          "__symbolic": "call",
          "expression": {
            "__symbolic": "reference",
            "module": "@angular/core",
            "name": "NgModule",
            "line": 2,
            "character": 1
          },
          "arguments": [
            {
              "declarations": [],
              "exports": [],
              "imports": [],
              "providers": []
            }
          ]
        }
      ],
      "members": {}
    }
  },
  "origins": {
    "AppModule": "./src/app/app.module",
    "ɵa": "./src/app/app.component",
    "ɵb": "./src/tools"
  },
  ...
}
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I don't believe this closes any open issue with this repo currently, although it is possibly a start for angular/angular#21082. However, I believe this is part of the root cause behind a range of issues in other libraries, including [ng-packagr](https://github.com/dherges/ng-packagr):

- dherges/ng-packagr#195
- dherges/ng-packagr#382

jeffora commented:
Seems like a test failure in Travis was due to running out of space - I don't think that has anything to do with this PR, as nothing here writes to disk.

jeffora commented:
It would be great if someone could get some eyes on this, even if just to say it's an unwanted change cc @matsko

hakimio commented:
@alxhub Would be really nice if you could review this pull request. It's a rather serious issue.

mhevery commented:
@jeffora I am sorry for the delay on getting back you. @alxhub is the best person to look at it and I have pinged him. In the meantime could you get the PR green, it is currently failing our CI.

alxhub commented on packages/compiler-cli/src/metadata/bundler.ts:
```diff
@@ -597,11 +597,20 @@ export class MetadataBundler {
+  getMetadataFor(fileName: string, containingFile: string): ModuleMetadata|undefined {
+    const {resolvedModule} =
+        ts.resolveModuleName(fileName, containingFile, this.options, this.host);
+    if (!resolvedModule) return undefined;
```
I think for safety, we should fall back on the old logic if `resolveModuleName` returns `undefined`. That is, we should append `.ts` to the file name in that case. That way we won't break existing apps if they somehow depend on our old broken behavior.

jeffora replied:
Updated this to fallback to old behavior if the module is not resolved.

alxhub reviewed (changes requested):
LGTM other than my one request. Thank you for doing this work!

alxhub commented:
http://test/OCL:203492592:BASE:203492621:1530897039126:ba045805

jeffora committed:
fix(compiler-cli): Use typescript to resolve modules for metadata

The current module resolution simply attaches .ts to the import/export path, which does
not work if the path is using Node / CommonJS behavior to resolve to an index.ts file.
This patch uses typescript's module resolution logic, and will attempt to load the original
typescript file if this resolution returns a .js or .d.ts file

jeffora commented:
@mhevery No problem! @alxhub I've updated to meet your changes, I'll keep an eye on the CI builds, but everything passed locally.

mhevery commented:
@jeffora It currently fails CI. Can you have a look.

admosity commented:
Some weird error? Is there some imposed way to import here?
```
packages/compiler-cli/test/metadata/bundler_spec.ts:15:59 - error TS2307: Cannot find module '../mocks'.

15 import {Directory, MockAotContext, MockCompilerHost} from '../mocks';
                                                             ~~~~~~~~~~
```
```
packages/compiler-cli/test/metadata/bundler_spec.ts
packages/compiler-cli/test/mocks.ts
```

jeffora commented:
@mhevery I've tried to recreate this locally, but I'm getting a bunch of different `bazel` errors. I've taken a guess at what I *think* the fix is and will see what happens with CI now. However, if it fails, some feedback or pointers for bazel would be great.

alxhub commented:
http://test/OCL:203492592:BASE:203841881:1531175473311:9624a123

jakeclarklucentmobile commented:
Sorry if this is a dumb question, need a little help.
How can I use this fix now that it has been approved? 
I currently have Angular Cli 6 installed via npm.

Do I need to pull this repo and build it manually or can I update my npm package somehow to obtain these changes?

alan-agius4 commented:
You should wait for the next Angular release which will include this change, and update your package.json version to point to the lastest once’s.

jakeclarklucentmobile commented:
Wonderful, do we have any idea when the next release will be cut? Is there a schedule?
Sorry again for silly questions. Thanks for your help

jasonaden commented:
@jakeclarklucentmobile It should be cut some time tomorrow morning. We generally cut a release once a week on Wednesdays.

ianchi commented:
@jeffora / @mhevery , I'm still having problems with barrel files.

I'm not 100% sure, but I think that this patch solves first level of barrel files, but the folder nesting is somehow lost, and when searching for nested imports in the resolved barrel, it treats the `folder_with_barrel` as a filename and it gets lost in the path for children:

Lets say:
`parent_folder/folder_with_barrel`

Gets correctly resolved, but the **relative imports** from inside `folder_with_barrel/index.ts`, are treated as relative to `parent_folder`. As if the `folder_with_barrel` part where a filename and not a folder.
And thus all those files are not found, and no metadata generated for them.

aburht commented:
Is this issue resolved?

aburht commented:
I'm still facing this issue with angular 7

reinzwei commented:
> @jeffora / @mhevery , I'm still having problems with barrel files.
> 
> I'm not 100% sure, but I think that this patch solves first level of barrel files, but the folder nesting is somehow lost, and when searching for nested imports in the resolved barrel, it treats the `folder_with_barrel` as a filename and it gets lost in the path for children:
> 
> Lets say:
> `parent_folder/folder_with_barrel`
> 
> Gets correctly resolved, but the **relative imports** from inside `folder_with_barrel/index.ts`, are treated as relative to `parent_folder`. As if the `folder_with_barrel` part where a filename and not a folder.
> And thus all those files are not found, and no metadata generated for them.

I am experiencing exactly same issue.
